alibi_detect/od/pytorch/svm.py

```diff
@@ -0,0 +1,441 @@
+
+        Parameters
+        ----------
+        fit_kwargs
+            dictionary of Kwargs to format. See `fit` method for details.
+
+        Returns
+        -------
+        Formatted kwargs.
+        """
+        return dict(
+            tol=fit_kwargs.get('tol', 1e-3),
+            max_iter=fit_kwargs.get('max_iter', 1000),
+            verbose=fit_kwargs.get('verbose', 0),
+        )
+
+    def score(self, x: torch.Tensor) -> torch.Tensor:
+        """Computes the score of `x`
+
+        Parameters
+        ----------
+        x
+            `torch.Tensor` with leading batch dimension.
+
+        Returns
+        -------
+        `torch.Tensor¬` of scores with leading batch dimension.
+
+        Raises
+        ------
+        NotFittedError
+            Raised if method called and detector has not been fit.
+        """
+        self.check_fitted()
+        x_nys = self.nystroem.transform(x)
+        x_nys = x_nys.cpu().numpy()
+        return self._to_tensor(-self.svm.score_samples(x_nys))
+
+
+class GdSVMTorch(SVMTorch):
+    ensemble = False
+
+    def __init__(
+        self,
+        nu: float,
+        kernel: 'torch.nn.Module',
+        n_components: Optional[int] = None,
+        device: Optional[Union[Literal['cuda', 'gpu', 'cpu'], 'torch.device']] = None,
+    ):
+        """Pytorch backend for the Support Vector Machine (SVM) outlier detector.
+
+        Parameters
+        ----------
+        nu
+            The proportion of the training data that should be considered outliers. Note that this does
+            not necessarily correspond to the false positive rate on test data, which is still defined when
+            calling the `infer_threshold` method.
+        kernel
+            Kernel function to use for outlier detection.
+        n_components
+            Number of components in the Nystroem approximation, by default uses all of them.
+        device
+            Device type used. The default tries to use the GPU and falls back on CPU if needed. Can be specified by
+            passing either ``'cuda'``, ``'gpu'``, ``'cpu'`` or an instance of ``torch.device``.
+        """
+
+        if (isinstance(device, str) and device == 'cpu') or \
+                (isinstance(device, torch.device) and device.type == 'cpu'):
+            warnings.warn(('The `gd` optimization option is best suited for GPU. If '
+                           'you want to use CPU, consider using the `sgd` option.'))
+
+        super().__init__(
+            device=device,
+            n_components=n_components,
+            kernel=kernel,
+            nu=nu,
+        )
+
+    def fit(  # type: ignore[override]
+        self,
+        x_ref: torch.Tensor,
+        step_size_range: Tuple[float, float] = (1e-6, 1.0),
+        n_step_sizes: int = 16,
+        tol: float = 1e-6,
+        n_iter_no_change: int = 25,
+        max_iter: int = 1000,
+        verbose: int = 0,
+    ) -> Dict:
+        """Fit the SVM detector.
```

Comment:
If we are being more descriptive about what is being fit in `SgdSVMTorch.fit` method, is it worth being slightly more descriptive here? i.e. `Fit the pytorch SVM detector` or similar?